EIP-005-extism-versioning-system.md

```diff
@@ -0,0 +1,79 @@
+2. There needs to be a simple way to understand what version of the runtime your dependency is compatible with.
+
+A naive solution would be to release all channels in the ecoysytem in lockstep. For example, when the runtime goes from 1.x.x to 2.x.x, all the host SDKs should go to 2.x.x and all the plugin PDKs should go to 2.x.x. The primary issue here iis that this does not communicate actual compatibilty changes. Furthermore, we will have instances where individual channels, e.g. Ruby Host SDK, will need breaking changes but the rest of the ecosystem does not. 
+
+## Solution
+
+The solution rests on two principles:
+
+1. We will use semver for each release channel. Each channel can be released indpendently of other components of the ecosystem. In this way, the semver version will only communicate to the programmer the relative changes in the channel, but not communicate compatibility with other parts of the ecoysystem.
+2. Compatibility will be determined by a 2-word, codeword system. Each major version release of the runtime will get a new codeword. The first word will indicate Host compatibility and the second word will indicate Plugin compatibility. The codeword will be of the form `<html-color>-<sea-creature>`.
+
+The key to understanding this is to look at the ecosystem as having two coupling points, or two "APIs". The Host API and the Plugin API:
+
+![libextism coupling](content/005-coupling.png)
+
+Changes in the Host API will break the SDK libraries but not the PDKs and vice versa. We may encounter contradicitons to this but we should hopefully be able to limit the changes to just the SDK. The Plugin API should rarely change as it breaks the entire world of built plugins. But the Host API may change more frequently. Thus we have decided to tie the version of the runtime to breaking changes in the Host API. Every new major bump in the runtime will represent a breaking change in the Host SDK. Rarely, a major bump in the runtime will indicate a breaking change in the Host OR the Plugin API. 
+
+The following is a hypothetical future state of the ecoystem:
+
+![Versioning Table](content/005-versioning.png)
+
+There are a few different methods we can explore to communicate compatibility. First I think we need documentation and to keep a table similar to the tables above. This should be generated from some parseable source of truth, like yaml.
+
+Second, I think we can encode into the SDK and PDK which versions they are compatible with. The way to do this would be to share a function b/w the Host API and the Plugin API. There is already a function which we could re-use in the host `const char *extism_version(void)`. We could also export this function into the Plugin API.
+
+The function would return a string like this `<semver-version>-<color>-<sea-creature>`. e.g.:
```

Comment:
This makes the whole string valid semver. Maybe we don't want to do that an express these are 2 different things by instead having a format like `<semver-version>;<color>-<edition>`

Reply:
its kind of nice that the whole string is semver, which has a side-effect of semver parsers being able to make sense of it?